sycl/tools/syclbin-dump/syclbin-dump.cpp

```diff
@@ -0,0 +1,193 @@
+//==----------- syclbin-dump.cpp -------------------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// The "syclbin-dump" utility lists the contents of a SYCLBIN file in a
+// human-readable format.
+//
+
+#include "llvm/Object/OffloadBinary.h"
+#include "llvm/Object/SYCLBIN.h"
+#include "llvm/Support/CommandLine.h"
+
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <regex>
+#include <string>
+
+using namespace llvm;
+
+thread_local size_t CurrentIndentationLevel = 0;
+
+class ScopedIndent {
+public:
+  ScopedIndent(size_t Indents = 2) : Incremented(Indents) {
+    CurrentIndentationLevel += Incremented;
+  }
+
+  ~ScopedIndent() { CurrentIndentationLevel -= Incremented; }
+
+  std::string str() const { return std::string(CurrentIndentationLevel, ' '); }
+
+private:
+  friend std::ostream &operator<<(std::ostream &OS, const ScopedIndent &IH);
+
+  const size_t Incremented;
+};
+
+std::ostream &operator<<(std::ostream &OS, const ScopedIndent &IH) {
+  return (OS << IH.str());
+}
+
+std::string_view StateToString(llvm::object::SYCLBIN::BundleState State) {
+  switch (State) {
+  case llvm::object::SYCLBIN::BundleState::Input:
+    return "input";
+  case llvm::object::SYCLBIN::BundleState::Object:
+    return "object";
+  case llvm::object::SYCLBIN::BundleState::Executable:
+    return "executable";
+  default:
+    return "UNKNOWN";
+  }
+}
+
+std::string_view IRTypeToString(llvm::object::SYCLBIN::IRType IRType) {
+  switch (IRType) {
+  case llvm::object::SYCLBIN::IRType::SPIRV:
+    return "SPIR-V";
+  case llvm::object::SYCLBIN::IRType::PTX:
+    return "PTX";
+  case llvm::object::SYCLBIN::IRType::AMDGCN:
+    return "AMDGCN";
+  default:
+    return "UNKNOWN";
+  }
+}
+
+std::string PropertyValueToString(const llvm::util::PropertyValue &PropVal) {
+  switch (PropVal.getType()) {
+  case llvm::util::PropertyValue::UINT32:
+    return std::to_string(PropVal.asUint32());
+  case llvm::util::PropertyValue::BYTE_ARRAY:
+    return std::string{PropVal.data(), PropVal.getByteArraySize()};
+  case llvm::util::PropertyValue::NONE:
+    break;
+  }
+  return "!UNKNOWN PROPERTY VALUE TYPE!";
+}
+
+void PrintProperties(llvm::util::PropertySetRegistry &Properties) {
+  for (auto PropertySet : Properties) {
+    ScopedIndent Ind;
+    std::cout << Ind << PropertySet.first.c_str() << ":\n";
+    for (auto PropertyValue : PropertySet.second) {
+      ScopedIndent Ind;
+      std::string PropValStr = PropertyValueToString(PropertyValue.second);
+      // If there is a newline in the value, start at next line and do
```

Comment:
should this function be owned by `llvm::util::PropertySetRegistry` instead?

Reply:
I am two minds regarding that. On one hand it could be useful elsewhere, but on the other hand this tool may know more about the actual properties inside the property set and could be adjust to handle specific property sets in a special way.

Reply:
I don't doubt having a `print` member function fits with the typical LLVM codebase, just like having a `dump` member function would be. However, I worry we lose control over formatting, like the indentation of these line-split strings, if we were to try and generalize it in a `print` method in `PropertySetRegistry`.